### ISSUE TITLE
Import schlägt fehl, wenn optionales Feld "publisherName" nicht vorhanden ist.

### ISSUE THREAD

**Issue by hofrob** [OPUS4/application#16]
XML:

```
<?xml version="1.0" encoding="UTF-8" ?>
<import xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="vendor/opus4-repo/framework/library/Opus/Util/opus_import.xsd">
    <opusDocument oldId="" language="eng" type="masterthesis" serverState="published">
        <titlesMain>
            <titleMain language="eng">
                A Generic Title For A Thesis
            </titleMain>
        </titlesMain>
        <persons>
            <person role="author" firstName="Person" lastName="A"/>
        </persons>
        <dates>
            <date type="published" year="2016"/>
        </dates>
    </opusDocument>
</import>
```

> Error while saving imported document # to database: SQLSTATE[HY000]: General error: 1364 Field 'publisher_name' doesn't have a default value, query was: INSERT INTO \`documents\` (\`type\`, \`language\`, \`published_year\`, \`server_date_created\`, \`server_date_modified\`, \`server_date_published\`, \`server_state\`) VALUES (?, ?, ?, ?, ?, ?, ?)

Kann leicht umgangen werden indem man das Attribut `publisherName` leer einfügt:

    <opusDocument oldId="" language="eng" type="masterthesis" serverState="published" publisherName="">

Getestet auf branch master und 4.5.

**Comment by hofrob** [comment]
Fehler tritt auch auf, wenn man bei einer frischen Installation folgende Schritte durchführt:

Klick auf "Veröffentlichen": https://example.com/publish
Irgendeine Dokumentart wählen
Checkbox "rights" anhaken
Klick auf "Weiter zum nächsten Schritt"

![screenshot from 2016-10-05 11-25-38](https://cloud.githubusercontent.com/assets/1058012/19107857/82425fca-8aee-11e6-83fd-20c4ddebdf52.png)

    SQLSTATE[HY000]: General error: 1364 Field 'publisher_name' doesn't have a default value, query was: INSERT INTO `documents` (`type`, `language`, `server_date_created`, `server_date_modified`, `server_state`) VALUES (?, ?, ?, ?, ?)

**Comment by hofrob** [comment]
Oh, habe gerade den Fix gesehen! Danke!!

Mein Quickfix für bestehende Installationen:

`ALTER TABLE opusdb.documents MODIFY publisher_name VARCHAR(255) COMMENT 'Name of an external publisher, e.g. Springer';`

**Comment by j3nsch** [comment]
Vielen Dank. Das Problem hat weniger mit dem Import als mit dem Datenbankzugriff im Allgemeinen zu tun. Verwenden Sie zufölligerweise Ubuntu 16. Bei unseren Tests mit dieser Ubuntu Version, die ich mittlerweile für die Entwicklung einsetze, ist uns aufgefallen, das MySQL mit anderen Defaults für den  SQL_MODE läuft. In den älteren Ubuntu Version (10-14) war der Defaultwert für den SQL_MODE anscheinend immer leer. Damit hat die Datenbank einfach "sinnvolle" Defaultwerte angenommen und diese Problem sind untergegangen. 

Ich bin dabei das Datenbankschema für den nächsten Release zu überarbeiten (4.5-Branch). Es sind leider wesentlich mehr Änderungen notwendig, teilweise auch im PHP Code, als nur der "publisher_name".  OPUS scheint unter Ubuntu 16 auf den ersten Blick zu laufen, aber es können Probleme beim Speichern von Objekten an ganz vielen Stelle auftreten. Mehrere hundert Unit Tests sind am Anfang gebrochen gewesen. Mittlerweile sind es wesentlich weniger und ich hoffe, dass wir bald 4.5 freigeben können.

**Comment by j3nsch** [comment]
Ich muss nachfragen. Welchen Fix?

Ich mach das Ticket mal wieder auf, weil es meiner Meinung nach noch für die offizielle Version besteht und bisher nur für den 4.5 Branch teilweise gefixt wurde.

**Comment by hofrob** [comment]
Ja, genau ich teste derzeit auf 16.04.1. Danke für die Infos!

In diesem [Commit](https://github.com/OPUS4/framework/commit/6a36d19a0cc2a6d4f5b0f23d3e63a1ab8f08f799) wurden die Defaults korrigiert. Diese Änderungen übernehme ich jetzt in meine DB. Kann gerne alle "ALTER"-Statements hier posten um das festzuhalten.

Der Import funktioniert, wenn man bei der Spalte das "NOT NULL" entfernt. Deswegen hab ich das Issue geschlossen :smile: 

lg, Robert

**Comment by j3nsch** [comment]
Stimmt. Jetzt bin ich durcheinander gekommen. Für das OPUS Framework, das hier hauptsächlich betroffen ist, arbeite ich momentan ohne 4.5 Branch, sondern aktualisiere MASTER direkt. Das muss ich irgendwann noch einmal durchdenken. Eine Installation des aktuellen Releases (MASTER) der Applikation, würde Version 4.5-RC1 des Frameworks verwenden und damit noch die Probleme haben. In der Entwicklungsversion arbeitet OPUS mit der neuesten Entwicklungsversion des Frameworks (dev-master auf Packagist.org). 

Sobald ich die Datenbankprobleme behoben habe, wird es auch ein Skript geben, um die notwendigen Änderungen an bestehenden Datenbanken vorzunehmen. 

Im Augenblick bin ich dabei Probleme im `Opus_DocumentFinder` (Datenbank-basierte Suche von Dokumenten) zu beheben, der SQL Queries generiert, die mit den SQL_MODE Einstellungen in Ubuntu 16 (die sicherlich sinnvoll sind für Produktionssysteme) nicht kompatibel sind.

**Comment by j3nsch** [comment]
Das Framework ist jetzt auf dem aktuellen Stand bei der Korrektur von bekannten Fehlern im Datenbankschema. 

Es gibt aber noch weitere Probleme, z.B. werden beim Editieren von Sammlungen NULL für Number oder OaiSubset als '0' angezeigt anstatt als leere Felder. Das schaue ich mir jetzt an.

@hofrob - Bitte gerne weitere Issues anlegen, wenn bei Ihren Tests unter Ubuntu 16 weitere Probleme auftauchen.

**Comment by j3nsch** [comment]
Das Editieren von Sammlungen mit NULL für Number oder OaiSubset ist doch in Ordnung.

**Comment by hofrob** [comment]
> Das Framework ist jetzt auf dem aktuellen Stand bei der Korrektur von bekannten Fehlern im Datenbankschema.

Super, danke schön!

> Bitte gerne weitere Issues anlegen, wenn bei Ihren Tests unter Ubuntu 16 weitere Probleme auftauchen.

Machen wir!
